File: Go/Path Sum.go
// 4ms, 97.23%; 4.8MB, 100.00%
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, sum int) bool {
	return root != nil && dfs(root, sum)
}

func dfs(root *TreeNode, target int) bool {
	if root.Left == root.Right {
		return target == root.Val
	}
	if root.Left != nil && dfs(root.Left, target-root.Val) {
		return true
	}
	if root.Right != nil && dfs(root.Right, target-root.Val) {
		return true
	}
	return false
}

File: Java/Path Sum.java
// 0ms, 100.00%; 39.2MB, 5.88%
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        return (null == root ? false : helper(root, sum));
    }

    private static boolean helper(TreeNode root, int sum) {
        int target = sum - root.val;
        if (root.left == root.right) {
            return 0 == target;
        }
        boolean flag1 = (null == root.left ? false : helper(root.left, target));
        boolean flag2 = (null == root.right ? false : helper(root.right, target));
        return flag1 || flag2;
    }
}
